snuba/datasets/errors_replacer.py

```diff
@@ -667,14 +665,126 @@ def process_unmerge(
         "previous_group_id": message["previous_group_id"],
         "project_id": message["project_id"],
         "timestamp": timestamp.strftime(DATETIME_FORMAT),
+        "hashes": _convert_hash_list(hashes, state_name),
     }
 
+    query_time_flags = (NEEDS_FINAL, message["project_id"])
+
+    return LegacyReplacement(
+        count_query_template, insert_query_template, query_args, query_time_flags
+    )
+
+
+def _convert_hash_list(
+    hashes: Sequence[str], state_name: ReplacerState, convert_types: bool = False
+) -> str:
     if state_name == ReplacerState.ERRORS:
-        query_args["hashes"] = ", ".join(
-            ["'%s'" % str(uuid.UUID(_hashify(h))) for h in hashes]
-        )
+        if convert_types:
+            return ", ".join(
+                "toUUID('%s')" % str(uuid.UUID(_hashify(h))) for h in hashes
+            )
+        else:
+            return ", ".join("'%s'" % str(uuid.UUID(_hashify(h))) for h in hashes)
     else:
-        query_args["hashes"] = ", ".join("'%s'" % _hashify(h) for h in hashes)
+        if convert_types:
+            return ", ".join("toFixedString('%s', 32)" % _hashify(h) for h in hashes)
+        else:
+            return ", ".join("'%s'" % _hashify(h) for h in hashes)
+
+
+def process_unmerge_hierarchical(
+    message: Mapping[str, Any],
+    all_columns: Sequence[FlattenedColumn],
+    state_name: ReplacerState,
+) -> Optional[Replacement]:
+    timestamp = datetime.strptime(message["datetime"], settings.PAYLOAD_DATETIME_FORMAT)
+    all_column_names = [c.escaped for c in all_columns]
+    select_columns = map(
+        lambda i: i if i != "group_id" else str(message["new_group_id"]),
+        all_column_names,
+    )
+
+    hierarchical_hashes = message.get("hierarchical_hashes") or {}
+    if not hierarchical_hashes:
+        return None
+
+    assert all(isinstance(h, str) for h in hierarchical_hashes.values())
+
+    prewhere = []
+    where = []
+    query_args = {}
+
+    # primary_hash is the root (most coarse) level of the grouping tree,
+    # meaning:
+    #
+    # 1. primary_hash == hierarchical_hashes[0]
+    # 2. <any entry in hierarchical_hashes>:primary_hash is a n:1 relation
+    #
+    # Sentry wants us to filter events by hierarchical_hashes entry using
+    # hasAny, but also tells us the corresponding primary_hash such that we
+    # can filter more efficiently.
+    #
+    # For now this is just converted into
+    #
+    #     PREWHERE primary_hash IN ... WHERE hasAny(hierarchical_hashes, ...)
+    #
+    # ..but perhaps one day we may decide to instead emit multiple replacements of the form
+    #
+    #     PREWHERE primary_hash = ... WHERE has(hierarchical_hashes, ...)
+    #
+    # ..if we think that it would be faster.
+    prewhere.append("primary_hash IN (%(primary_hashes_for_hierarchical_hashes)s)")
```

Comment:
Why not just calling this `primary_hashes_for_hierarchical_hashes`? It is the only primary hash we are dealing with in this method. No need to specify that it is the one associated with the hierarchical hashes.

Reply:
Also when would we ever have multiple different primary_hash here? If we are unmerging (splitting) Aren't all the groups we are splitting guaranteed to have the same primary_hash ?

Reply:
You are right, the constraint of not allowing to split groups with 2+ primary_hashes has been added at a later point so we can make the message more strict.